Make naked functions incompatible with certain attributes

bstrie opened the pull request:
See the comment thread beginning at https://github.com/rust-lang/rust/issues/90957#issuecomment-1028351639 for justification.

Commit breakdown:

1. **Fix copy/paste errors in check_attr**

2. **Consolidate checking for attributes not compatible with #[naked]**

    Currently #[inline] and #[track_caller] are incompatible with #[naked], and there are more such attributes on the way.

    In preparation for this, consolidate the existing checks into a central location so that they can all be handled together.

    In addition, clean up the wording of the existing error messages.

3. **Forbid #[target_feature] from being used on naked functions**

    Per the comment at https://github.com/rust-lang/rust/issues/90957#issuecomment-1029483354 , make these two attributes incompatible out of an abundance of caution. This may be allowed in the future.

4. **Forbid testing attributes from being used with #[naked]**

    Per the comment at https://github.com/rust-lang/rust/issues/90957#issuecomment-1029483354 , this prevents #[naked] from being used with #[test], #[should_panic], #[ignore], and #[bench].

    This also fixes an oversight where the #[naked]-related lint "undefined_naked_function_abi" was not being properly registered; it is now possible to allow/deny it as per usual.

bstrie committed:
Fix copy/paste errors in check_attr

rust-highfive commented:
Some changes occurred in diagnostic error codes

cc @GuillaumeGomez

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

GuillaumeGomez reviewed (approved):
Error explanations look good for me (they don't need a "fixed" example in this case considering how simple they are). Please wait for approval from someone on the compiler team before r+.

bstrie commented:
@GuillaumeGomez thanks, these were so simple that I wasn't even sure if they deserved their own extended explanations. Is it policy that every error should ideally have its own extended explanation someday? If not, what's the cutoff point for when not to make an extended explanation?

GuillaumeGomez commented:
If there is an error with an error code, then it should have an explanation (easy to remember :laughing:). There is a [tracking issue](https://github.com/rust-lang/rust/issues/61137) for this as well but it might need to be updated (new error codes have been added I suppose).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5126090291?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ef7273f5210ff0c8d37e8b2770a787e4d0644bd2/checks/5126090291/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
a different dependency.

Run `npm audit` for details.
npm notice 
npm notice New major version of npm available! 7.21.1 -> 8.4.1
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.4.1>
npm notice Run `npm install -g npm@8.4.1` to update!
Removing intermediate container 9d2c8a907b8c
 ---> b8b758ff971c
Step 6/12 : RUN npm install eslint@8.6.0 -g
 ---> Running in 157eaa91a13f
---
Successfully built 3dd5ba0bd0c0
Successfully tagged rust-ci:latest
Built container sha256:3dd5ba0bd0c0df1459a8be1a9f75cf8db73e6f998d2dbc37d6f267b3eb2a34cb
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 632 error codes
* highest error code: E0790
tidy error: /checkout/src/test/ui/asm/naked-functions-target-feature.rs: too many trailing newlines (2)
`/checkout/compiler/rustc_error_codes/src/error_codes/E0789.md` doesn't use its own error code in compile_fail example
Done!
* 363 features
some tidy checks failed
Build completed unsuccessfully in 0:00:12
```

</details>

bstrie commented:
@GuillaumeGomez right, but I have the same problem when deciding if an error is too simple to deserve its own error code. :P 

Also, I could use your advice in resolving the build failure. The problem is the doctest for one of the extended descriptions:
```
    ```compile_fail
    #[target_feature(enable = "sse2")]
    #[naked]
    unsafe fn foo() {}
    ```
```
The problem is that I'm using `compile_fail` instead of `compile_fail,E0789`. This is deliberate, because target_feature will fail to compile with a different error code when used on a platform that doesn't support sse2. How do I deal with platform-specific tests in this context?

jackh726 commented:
I could imagine in the future more attributes falling into this category.

bstrie commented:
> Would it better to have a single error code for invalid attribute mixing? (Or maybe just X + naked?)

Possibly, I'm open to suggestions as to what is most appropriate here. Alternatively we could have just two error codes, one for codegen attributes + naked and one for testing attributes + naked.

bstrie commented:
Thinking about it, I like the idea of consolidating these error codes into just "codegen attributes + naked" and "testing attributes + naked". As a bonus, we can avoid having an explicit example for target_feature, which sidesteps the need to bend over backwards to make this doctest platform-specific.

GuillaumeGomez commented:
> The problem is that I'm using `compile_fail` instead of `compile_fail,E0789`. This is deliberate, because target_feature will fail to compile with a different error code when used on a platform that doesn't support sse2. How do I deal with platform-specific tests in this context?

You can use `ignore` with parens explaining why and to add it into the tidy checker.

bstrie committed:
Consolidate checking for attributes not compatible with #[naked]

Currently #[inline] and #[track_caller] are incompatible with #[naked],
and there are more such attributes on the way.

In preparation for this, consolidate the existing checks into a central
location so that they can all be handled together.

In addition, clean up the wording of the existing error messages.

bstrie committed:
Forbid #[target_feature] from being used on naked functions

Per the comment at https://github.com/rust-lang/rust/issues/90957#issuecomment-1029483354
make these two attributes incompatible out of an abundance of caution.
This may be allowed in the future.

bstrie committed:
Forbid testing attributes from being used with #[naked]

Per the comment at https://github.com/rust-lang/rust/issues/90957#issuecomment-1029483354 ,
this prevents #[naked] from being used with #[test], #[should_panic], #[ignore],
and #[bench].

This also fixes an oversight where the #[naked]-related lint
"undefined_naked_function_abi" was not being
properly registered; it is now possible to allow/deny it as per usual.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5130617194?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/07b9b653dbe8e5eda9510d015bc6be4b93921079/checks/5130617194/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 3e9eebdf47e6
Successfully tagged rust-ci:latest
Built container sha256:3e9eebdf47e6266d32ab29b2aa7bbda2d1c7ee659dbf23b09a25ea4371d05b00
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 630 error codes
* highest error code: E0788
tidy error: /checkout/src/test/ui/asm/naked-functions-target-feature.rs: too many trailing newlines (2)
Found 0 error codes with no tests
Done!
* 363 features
some tidy checks failed
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5130851259?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e93ec1cc7abaa49c559a82e5235fb6c9cb09c71b/checks/5130851259/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built ea338b340e32
Successfully tagged rust-ci:latest
Built container sha256:ea338b340e32dbbebcbd8e6c2e8ecb2ed4b66bdd6523841a901105b357e784c8
Uploading finished image to https://ci-caches.rust-lang.org/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0
upload failed: - to s3://rust-lang-ci-sccache2/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0 Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-llvm-12]
---
Testing error-index stage2
doc tests for: /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md


command did not execute successfully: "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" ""

stdout ----

running 1036 tests
---
test /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0788 (line 16721) - compile fail ... FAILED

failures:

---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0788 (line 16721) stdout ----
Test compiled successfully, but it's marked `compile_fail`.
failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0788 (line 16721)

test result: FAILED. 1004 passed; 1 failed; 31 ignored; 0 measured; 0 filtered out; finished in 8.71s
```

</details>

npmccallum commented:
@bstrie Would it be better to have an allow list for acceptable naked functions attributes? This would deny all new attributes without any manual intervention.

bstrie commented:
> Would it be better to have an allow list for acceptable naked functions attributes? This would deny all new attributes without any manual intervention.

I think that would be an unnecessary maintenance burden. IMO, when a feature is stabilized, it is the responsibility of the people behind that feature to ensure that it works with all existing language features. Thus, with naked functions it is our responsibility to ensure that it works correctly with all existing attributes, and to prevent it from being used with any attributes that it is incompatible with. Then, once naked functions are stabilized, it is the responsibility of future language features (e.g. new attributes) to ensure that they work correctly with all existing language features, which will include naked functions. Unless someone wants to argue why naked functions represent some greater risk than usual, I consider it outside the scope of this stabilization effort to try to pre-emptively prevent problems that may possibly arise from interactions with an unknowable list of future features.

jackh726 commented:
I still am not sure if it makes sense to have *two* separate errors (codegen + testing). And even, if it's worth having a single error code for all "invalid attribute combinations". I think I'd feel more comfortable looping in either @rust-lang/wg-diagnostics or @rust-lang/compiler for thoughts.

bstrie commented:
Yes, to reiterate, I'm happy to have guidance on when it makes sense to use an error code and when it makes sense to just produce an anonymous error. Ultimately the decision of how many error codes should exist here is relatively minor in the grand scheme of this PR compared to the rejection logic and the tests.

npmccallum commented:
> I think that would be an unnecessary maintenance burden. IMO, when a feature is stabilized, it is the responsibility of the people behind that feature to ensure that it works with all existing language features.

How do we know that new feature authors have done this due diligence regarding naked functions? They add their new feature to the allow list with appropriate test coverage.

All new features would be opt-in. So at least we wouldn't break existing code. However, where people do opt in, the failures can be subtle and hard to debug. Even worse, they might not fail but instead silently succeed with security implications (i.e. the stack is corrupted).

I'm on the fence on an allowlist, but leaning towards pro.

bstrie commented:
To ease any decision regarding error codes, here's a summary of the situation:

1. `#[naked]` is currently incompatible with `#[inline]` and `#[track_caller]`.
2. Using `#[naked]` and `#[track_caller]` together produces an error with a code, whereas using `#[naked]` together with `#[inline]` produces an error with no code.
3. `#[naked]` needs to also reject `#[target_feature]` and `#[test]`, which is implemented in this PR.
4. The first version of this PR made every combination of `#[naked]` and these four attributes into its own unique error code, making for four error codes total.
5. As of this writing, the current version of this PR consolidates `#[inline]`, `#[track_caller]`, and `#[target_feature]` into a single error code for "codegen attributes incompatible with `#[naked]`, while `#[test]` still has its own error code; two error codes total.
6. Other alternatives:
    * Have no error codes for any of these errors.
    * Have one broad error code for **all** errors of the form "you used two attributes that were incompatible" (it's possible that this already exists)
    * Have one error code solely for errors of the form "you used `#[naked]` with an incompatible attribute"

bstrie commented:
> How do we know that new feature authors have done this due diligence regarding naked functions?

It is assumed that the lang team will hold feature authors to this standard before allowing a feature to be stabilized. It's not impossible for things to slip through--accidents happen--but this is the acceptable risk that comes with stabilizing any feature.

> Even worse, they might not fail but instead silently succeed with security implications (i.e. the stack is corrupted).

The threat in question involves a future attribute `#[foo]` that is silently incompatible with naked functions. The proposed mitigation is to prevent all but a hand-selected list of attributes on naked functions. However, for this threat to be realized the author of the naked function would have to manually add `#[foo]` to the naked function; every other existing user of naked functions cannot possibly be affected (at least not by any mechanism that the allowlist proposal would be capable of preventing). Thus, it would be impossible to realize such a threat merely by upgrading one's toolchain; you would have to first upgrade your toolchain, and then consciously adopt `#[foo]` into your codebase in order to be affected. Finally, `#[foo]` would have to be an attribute that affects codegen somehow, and (correct me if I'm wrong) it seems reasonable to assume that any security-minded user will have integration tests in place that would detect a failure on the PR that adds `#[foo]` to their naked functions. Thus IMO the benefit here is small.

Meanwhile, the cost is very large. I find it eminently reasonable to assume that a feature author that adds an attribute that affects codegen will perform their due diligence in remembering to deny its use on `#[naked]` (especially since they will be forced to update the reference page on codegen attributes, which will by then contain the documentation on attributes that may not be used with naked). I also find it reasonable to assume that feature authors that add an attribute that does *not* affect codegen will reasonably assume that they do not need to arbitrarily allow their attribute on naked functions; I cannot think of any other precedent for this in the language. What seems unreasonable to me would be forcing these latter authors to remember to manually allow their attribute to be used with naked functions; in practice I suspect they  would simply not, and users of naked functions would be frustrated by the inability to do things that should obviously be allowed. Between the burden on feature authors and the burden on users who would be prevented from doing things for no reason (especially when considering that this policy would also prevent third-party proc macros from being used with naked functions), I feel that the cost outweighs the benefit, IMO.

npmccallum commented:
> > How do we know that new feature authors have done this due diligence regarding naked functions?
> 
> It is assumed that the lang team will hold feature authors to this standard before allowing a feature to be stabilized. It's not impossible for things to slip through--accidents happen--but this is the acceptable risk that comes with stabilizing any feature.
> 
> > Even worse, they might not fail but instead silently succeed with security implications (i.e. the stack is corrupted).
> 
> The threat in question involves a future attribute `#[foo]` that is silently incompatible with naked functions. The proposed mitigation is to prevent all but a hand-selected list of attributes on naked functions. However, for this threat to be realized the author of the naked function would have to manually add `#[foo]` to the naked function; every other existing user of naked functions cannot possibly be affected (at least not by any mechanism that the allowlist proposal would be capable of preventing). Thus, it would be impossible to realize such a threat merely by upgrading one's toolchain; you would have to first upgrade your toolchain, and then consciously adopt `#[foo]` into your codebase in order to be affected. Finally, `#[foo]` would have to be an attribute that affects codegen somehow,

I agree up until this point.

> and (correct me if I'm wrong) it seems reasonable to assume that any security-minded user will have integration tests in place that would detect a failure on the PR that adds `#[foo]` to their naked functions. Thus IMO the benefit here is small.

This is a very bad assumption. All users are impacted by these problems regardless if they are security-minded or not. And the interactions between these attributes are non-obvious.

> Meanwhile, the cost is very large.

Given that the number of users that consumes Rust features is exponentially larger than the number of Rust feature developers, the total cost is significantly smaller to the first group than to the second.

> I find it eminently reasonable to assume that a feature author that adds an attribute that affects codegen will perform their due diligence in remembering to deny its use on `#[naked]` (especially since they will be forced to update the reference page on codegen attributes, which will by then contain the documentation on attributes that may not be used with naked).

I would hope so indeed. But when this fails, it fails spectacularly. Trying to bisect this will be very painful. OTOH, being able to do a `git blame` against an allowlist gives us the relevant commit(s) immediately.

> I also find it reasonable to assume that feature authors that add an attribute that does _not_ affect codegen will reasonably assume that they do not need to arbitrarily allow their attribute on naked functions; I cannot think of any other precedent for this in the language.

This is because for other features that are this dangerous, like `asm`, you don't just automatically get integration. For example, if you want to pass a new kind of value to `asm` you have to explicitly allow support for it in the `asm` functionality. No built-in hurdle like this exists for naked functions. The problem here is, I think, unique because we have something potentially very dangerous that exists precisely at the junction of an integration point (function attributes) where the number of combinations is high and difficult to reason about.

> What seems unreasonable to me would be forcing these latter authors to remember to manually allow their attribute to be used with naked functions; in practice I suspect they would simply not, and users of naked functions would be frustrated by the inability to do things that should obviously be allowed.

But... doing this is safe. And fixing the oversight is trivial if there are no safety concerns. If you stabilize your feature and realize you've forgotten to add it to the allowlist, this should be a trivial patch that can be merged within the next release window. It isn't like you have to go back to the drawing board.

It seems to me like you're arguing that it is reasonable to expect feature developers to remember to consider all the subtle interactions between codegen attributes but not reasonable to expect the same feature developers to remember to add one item to an allowlist. The former is far more work than the latter.

> Between the burden on feature authors and the burden on users who would be prevented from doing things for no reason (especially when considering that this policy would also prevent third-party proc macros from being used with naked functions), I feel that the cost outweighs the benefit, IMO.

It seems to me like you're optimizing for the burden of the feature authors rather than for the safety of the consumers of this feature.

bstrie commented:
> And the interactions between these attributes are non-obvious.

Certainly, which is part of the problem with concerning ourselves with general hypotheticals. It would be easier to discuss the likely threat of this scenario if we had a more concrete example.

> Given that the number of users that consumes Rust features is exponentially larger than the number of Rust feature developers, the total cost is significantly smaller to the first group than to the second.

As mentioned, users looking to use attributes on their naked functions would also be bearing this cost. This is especially true for users of custom proc macros, since Rust cannot add external proc macros to the allowlist.

> It seems to me like you're arguing that it is reasonable to expect feature developers to remember to consider all the subtle interactions between codegen attributes but not reasonable to expect the same feature developers to remember to add one item to an allowlist.

Rather, Rust's orthogonality usually allows feature developers to ignore most of Rust's features when implementing new features. The only existing features that need to be considered are those that have some likely overlap with the new feature; this quality makes new feature development much more tractable than it would be otherwise. This is why I suggest that it is likely that the author of a new codegen attribute (which will plausibly interact poorly with `#[naked]`) is likely to consider its interaction with `#[naked]`, and why the author of a non-codegen attribute (which will not interact poorly with `#[naked]`) is not likely to consider its interaction with `#[naked]`.

> It seems to me like you're optimizing for the burden of the feature authors rather than for the safety of the consumers of this feature.

I suspect that we may have gotten sidetracked with the safety argument here. The reason that the attributes in this PR are being made incompatible with `#[naked]` is not for safety, but for future compatibility. The goal AIUI is to be able to leave the door open for a future version of Rust to have the option to lower a naked function to a semantically-equivalent `global_asm!` invocation. But even for safety, I consider the allowlist approach less than stellar; I am less concerned with future developments adding a `#[foo]` attribute incompatible with naked functions (which the allowlist approach would protect against) than I am with future developments making e.g. the `#[doc]` attribute incompatible with naked functions (which the allowlist approach would not protect against, and would result in vulnerabilities merely from updating one's toolchain).

Despite the amount of spilled ink here, I actually don't care all that much which approach is taken. My primary concern is with getting naked functions stabilized ASAP, and for the immediate future the allowlist approach and the blocklist approach will be semantically identical, and there's nothing stopping a future version of Rust from switching from one to the other. For maximum expediency, I will bring up to the lang team and implement whichever they currently think is the best approach.

Aaron1011 commented:
> This is especially true for users of custom proc macros, since Rust cannot add external proc macros to the allowlist.

Custom proc macro attributes are all expanded (and removed) early in compilation. The only attributes remaining after macro expansion are built-in attributes, so the allowlist would only need to consider built-in attributes

nikomatsakis commented:
I was reading the back and forth, @npmccallum and @bstrie, but I'm feeling a bit lost. @npmccallum, are you arguing *against* an allow-list? Do you have an alternative proposal?

bstrie commented:
@nikomatsakis , @npmccallum is arguing that the compiler should have a list of attributes that are allowed to be present on `#[naked]`, and should reject any attribute not on the list. That would differ from what this PR currently implements, which is a list of attributes that should be rejected from appearing with `#[naked]`, where any unlisted attributes are allowed.

jackh726 commented:
Status:
- Needs rebase
- Needs change to "allow list"
- Need clarification on how we want error codes to look (share error code with others, one error code, or multiple error codes)

Dylan-DPC commented:
@bstrie any updates on this?

estebank commented on src/test/ui/asm/naked-functions-testattrs.stderr:
```diff
@@ -0,0 +1,27 @@
+error[E0788]: cannot use testing attributes with `#[naked]`
+  --> $DIR/naked-functions-testattrs.rs:12:1
+   |
+LL | #[naked]
+   | ^^^^^^^^
```
In the output, can we also point with secondary spans to the other non applicable attributes?

estebank commented on src/test/ui/asm/naked-functions-target-feature.stderr:
```diff
@@ -0,0 +1,9 @@
+error[E0736]: cannot use additional code generation attributes with `#[naked]`
+  --> $DIR/naked-functions-target-feature.rs:21:1
+   |
+LL | #[target_feature(enable = "sse2")]
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Can we point at `#[naked]` as well here, and *maybe* at the item name too?

JohnCSimon commented:
@bstrie  
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
